### ISSUE TITLE
 Cypress executable not found at (on Vercel installation)

### ISSUE THREAD

**Issue by Olivier-Vromans** [cypress-io/cypress#24406]
### Current behavior

When I try to download cypress as npm package on Vercel I get that installation success but when I try to open Cypress with the api (Can only use that) it says: "Cypress executable not found"

```
2022-10-26T14:28:09.186Z  Installing Cypress (version: 10.10.0)
2022-10-26T14:28:09.186Z  
2022-10-26T14:28:09.192Z  [STARTED] Task without title.
2022-10-26T14:28:11.954Z  [SUCCESS] Task without title.
2022-10-26T14:28:11.955Z  [STARTED] Task without title.
2022-10-26T14:28:21.710Z  [SUCCESS] Task without title.
2022-10-26T14:28:21.710Z  [STARTED] Task without title.
2022-10-26T14:28:21.745Z  [SUCCESS] Task without title.
2022-10-26T14:28:22.747Z  
2022-10-26T14:28:22.748Z  You can now open Cypress by running: node_modules/.bin/cypress open
2022-10-26T14:28:22.748Z  
2022-10-26T14:28:22.748Z  https://on.cypress.io/installing-cypress
2022-10-26T14:28:22.748Z  

2022-10-26T14:30:13.446Z  Please reinstall Cypress by running: cypress install
2022-10-26T14:30:13.446Z  
2022-10-26T14:30:13.446Z  ----------
2022-10-26T14:30:13.446Z  
2022-10-26T14:30:13.446Z  Cypress executable not found at: /home/sbx_user1051/.cache/Cypress/10.10.0/Cypress/Cypress
2022-10-26T14:30:13.446Z  
2022-10-26T14:30:13.446Z  ----------
2022-10-26T14:30:13.446Z  
2022-10-26T14:30:13.446Z  Platform: linux-x64 (Amazon Linux - 2 (Karoo))
2022-10-26T14:30:13.446Z  Cypress Version: 10.10.0
2022-10-26T14:30:13.446Z      at /var/task/node_modules/cypress/lib/errors.js:348:17 {
2022-10-26T14:30:13.446Z    known: true
2022-10-26T14:30:13.446Z  }

```

### Cypress Version

10.10.0

### Node version

v16.x

### Package Manager

npm

### Package Manager Version

v7 (found that online)

### Operating system

Linux

### Operating System Version

linux-x64 (Amazon Linux - 2 (Karoo))

**Comment by Olivier-Vromans** [comment]
I have also tried installing it as a dev-dependecie but than the build fails and using these methodes:
```
Running "install" command: `npm install && node_modules/.bin/cypress run`...

[STARTED] Task without title.
[FAILED] Your system is missing the dependency: Xvfb
[FAILED] 
[FAILED] Install Xvfb and run Cypress again.
[FAILED] 
[FAILED] Read our documentation on dependencies for more information:
[FAILED] 
[FAILED] https://on.cypress.io/required-dependencies
[FAILED] 
[FAILED] If you are using Docker, we provide containers with all required dependencies installed.
[FAILED] 
[FAILED] ----------
[FAILED] 
[FAILED] Error: spawn Xvfb ENOENT
[FAILED] 
[FAILED] ----------
[FAILED] 
[FAILED] Platform: linux-x64 (Amazon Linux - 2 (Karoo))
[FAILED] Cypress Version: 10.11.0
Your system is missing the dependency: Xvfb
Install Xvfb and run Cypress again.
Read our documentation on dependencies for more information:
https://on.cypress.io/required-dependencies
If you are using Docker, we provide containers with all required dependencies installed.
----------
Error: spawn Xvfb ENOENT
----------
```

```
Running "install" command: `npm install && npm install xvfb && node_modules/.bin/cypress run`...

[STARTED] Task without title.
--
[FAILED] Your system is missing the dependency: Xvfb
[FAILED]
[FAILED] Install Xvfb and run Cypress again.
[FAILED]
[FAILED] Read our documentation on dependencies for more information:
[FAILED]
[FAILED] https://on.cypress.io/required-dependencies
[FAILED]
[FAILED] If you are using Docker, we provide containers with all required dependencies installed.
[FAILED]
[FAILED] ----------
[FAILED]
[FAILED] Error: spawn Xvfb ENOENT
[FAILED]
[FAILED] ----------
[FAILED]
[FAILED] Platform: linux-x64 (Amazon Linux - 2 (Karoo))
[FAILED] Cypress Version: 10.11.0
Your system is missing the dependency: Xvfb

Install Xvfb and run Cypress again.
 
Read our documentation on dependencies for more information:
https://on.cypress.io/required-dependencies
 
If you are using Docker, we provide containers with all required dependencies installed.
---------- 
Error: spawn Xvfb ENOENT
----------
```

**Comment by amehta265** [comment]
@Olivier-Vromans Thank you for posting this issue.

I was able to start a simple `vercel` project and use the Cypress `module api` to run a simple e2e test.
It would be more helpful if you can provide your current `cypress.config` file, your `package.json` and maybe a screenshot of the test you are trying to run so that I can more closely mimic your set up and reproduce this error.

If you can provide a [Short, Self Contained, Correct, Example](http://sscce.org/) project that contains the error you are facing that would be even better!

Furthermore, as mentioned in your error logs, have you tried downloading `Xvfb` on your linux machine. It seems like you may not have this dependency which could be the reason behind this error. This is a required dependency for running Cypress on a linux based system. Cypress usually spawns its own `X11 server` but if you are running several cypress instances in parallel, you make need to use set a `display` attribute. You can get more information here about Xvfb and Cypress (https://docs.cypress.io/guides/continuous-integration/introduction#In-Docker)

**Comment by Olivier-Vromans** [comment]
@amehta265 Thank you for responding, I tried to start a simple Vercel project with cypress with a simple e2e test too but no luck. These were my steps:

I created a new Nextjs project in Vercel by using the template. Then I cloned the project to my computer and openend it with VSCode. I added cypress by using the `npm install cypress --save-dev` command and added the `cypress:open/run` scripts to the `package.json` file. After that I openend cypress and initiated cypress. I added the code that I use to the project en the test was working **localy**. I pushed than the code to github so Vercel would build. But no luck again it said cypress was not found.

My repo I used to for the recreation: [Cypress-test](https://github.com/Olivier-Vromans/Cypress-test).

this is my `package.json`:

```
{
  "name": "nextjs",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "cypress:open": "cypress open",
    "cypress:run": "cypress run"
  },
  "dependencies": {
    "next": "13.0.0",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "zustand": "^4.1.3"
  },
  "devDependencies": {
    "cypress": "^10.11.0",
    "eslint": "8.26.0",
    "eslint-config-next": "13.0.0"
  }
}
```


and this is my `cypress.config`:
```
const { defineConfig } = require("cypress");

module.exports = defineConfig({
  e2e: {
    video: false,
    "retries": {
      // Configure retry attempts for `cypress run`
      // Default is 0
      "runMode": 1
    },
    trashAssetsBeforeRuns: true,
    setupNodeEvents(on, config) {
      // implement node event listeners here
    },
  },
});
```

This is the `execute.js` file that I use to call the cypress run in
```
import cypress from 'cypress';

export default async function handler(req, res) {
  if (req.method === "POST") {
    // Get selected sites from request body
    const args = req.body
    let spec = []
    let screens = []
    if (args.length === 0) {
      console.log("No parameters");
      // push specific spec file in a array to run those tests
      spec.push('./cypress/e2e')
    } else {
      console.log("Parameters found");
      for (const arg of args) {
        spec.push('./cypress/e2e/' + arg)
        // Screenshots will be saved in cypress/screenshots folder with the name of the spec file
        screens.push('./cypress/screenshots/' + arg + "/")
      }
    }

    if (spec.length !== 0) {
      console.log('Run cypres for: ' + args);
      console.log('Screenshot directories: ' + screens);
      cypress.run({
        // the path is relative to the current working directory
        spec: spec,
      }).then((results) => {
        console.log(results);
        return res.status(200).json({
          status: "Completed"
        });
      })
        .catch((err) => {
          console.error(err)
        })
    } else {
      return null
    }
  }
}

```


this is my `kav.cy.js` test script I tried to use:
```
/// <reference types="Cypress" />

beforeEach("https://kavautoverhuur.nl/",() => {
    cy.visit("https://kavautoverhuur.nl/")        
})

describe("Homepage of KAV Auto Verhuur", () => {
    it("Catch empty a tags", () => {
        cy.get("a[href]")
            .each($el => {
                cy.wrap($el).should('have.attr', 'href').should('not.be.empty')
            })
    })
})

```


log from vercel:

```
Cloning github.com/Olivier-Vromans/Cypress-test (Branch: main, Commit: df27ad2)

Cloning completed: 658.377ms
Looking up build cache...
Build cache downloaded [70.85 MB]: 4113ms
Running "vercel build"
Vercel CLI 28.4.12-05a80a4
Installing dependencies...

up to date in 877ms

101 packages are looking for funding
run `npm fund` for details
Detected Next.js version: 13.0.0
Detected `package-lock.json` generated by npm 7+...
Running "npm run build"

> nextjs@0.1.0 build
> next build

info  - Linting and checking validity of types...
info  - Creating an optimized production build...
info  - Compiled successfully
info  - Collecting page data...
info  - Generating static pages (0/3)
info  - Generating static pages (3/3)
info  - Finalizing page optimization...

Route (pages)                              Size     First Load JS
┌ ○ /                                      1.94 kB        74.1 kB
├   └ css/ae0e3e027412e072.css             707 B
├   /_app                                  0 B            72.1 kB
├ ○ /404                                   181 B          72.3 kB
├ λ /api/execute                           0 B            72.1 kB
└ λ /api/hello                             0 B            72.1 kB
+ First Load JS shared by all              72.4 kB
  ├ chunks/framework-8c5acb0054140387.js   45.4 kB
  ├ chunks/main-d2a793b6dc23a82a.js        25.7 kB
  ├ chunks/pages/_app-b982fa09370b0414.js  284 B
  ├ chunks/webpack-8fa1640cc84ba8fe.js     750 B
  └ css/ab44ce7add5c3d11.css               247 B

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)

Traced Next.js server files in: 1.291s
Created all serverless functions in: 1.024s
Collected static files (public/, static/, .next/static): 10.183ms
Build Completed in /vercel/output [14s]
Generated build outputs:
 - Static files: 18
 - Serverless Functions: 2
 - Edge Functions: 0
Serverless regions: Washington, D.C., USA
Deployed outputs in 2s

Build completed. Populating build cache...
Uploading build cache [70.84 MB]...
Build cache uploaded: 1.668s
Done with "."
INFO	No parameters
INFO	Run cypres for: 
INFO	Screenshot directories: 
ERROR	Error: No version of Cypress is installed in: /home/sbx_user1051/.cache/Cypress/10.11.0/Cypress

Please reinstall Cypress by running: cypress install

----------

Cypress executable not found at: /home/sbx_user1051/.cache/Cypress/10.11.0/Cypress/Cypress

----------

Platform: linux-x64 (Amazon Linux - 2 (Karoo))
Cypress Version: 10.11.0
2022-10-28T08:07:29.754Z      at /var/task/node_modules/cypress/lib/errors.js:348:17 {
2022-10-28T08:07:29.754Z    known: true
2022-10-28T08:07:29.754Z  }
Task timed out after 10.01 seconds
```

**Comment by amehta265** [comment]
@Olivier-Vromans Thank you for providing detailed steps for creating a NextJS project in Vercel using the template as well as providing your reproducible example.
I managed to clone your repository and run cypress successfully locally (as you mentioned). I was also able to `npm run buid` on your project and the build was successful and no errors were thrown.

I have gone a step further and have created a `nextJs` project from scratch on Vercel and have deployed it to my github. It is an exact copy of your project and contains the same dependencies and files. You can check it out here: https://github.com/amehta265/nextjs. Unfortunately everything seems to be working smoothly on this new project as well.

The reason for this issue may be that cypress tries to access the `cache` when looking for Cypress libraries. However, something on your machine maybe clearing the cache causing cypress to fail


At this point kindly try one of the following things:
- Try to [clear cypress cache](https://docs.cypress.io/guides/guides/command-line#cypress-cache-clear) and then re-install cypress `npm install cypress --save-dev`
- Run  `npm install cypress --save-dev` and then run  `./node_modules/.bin/cypress install`. You have to wait to the installation is completed and after that you will be able to run the open command: `node_modules/.bin/cypress open`
- Try to run this:
```
> rm -rf ~/.cache/Cypress/<version-from-the-error-here>/
> npm cache clear -f
> npm i cypress -D
> node_modules/.bin/cypress install
```

I hope one of these solutions works for you. For added information and to see what worked for other users check out this [related issue](https://github.com/cypress-io/cypress/issues/2610)

**Comment by Olivier-Vromans** [comment]
Unfortunately is that also not working in vercel for me. I also tried the methodes from the related issue but that isn't working for me either. Vercel keeps saying cypress isn't installed but I found out that it is getting installed in `/vercel/.cache/Cypress/10.11.0` and try to run it from `/home/sbx_user1051/.cache/Cypress/10.11.0/Cypress/Cypress` so maybe that is the issue but I have no Idea how I could change so it would look in the right path? Also your repository isn't available, when I try and open it I get a 404 page not found

**Comment by amehta265** [comment]
@Olivier-Vromans I have changed the visibility settings for the repository. You should be able to access it now. Kindly let me know if this is not the case: https://github.com/amehta265/nextjs

I am looking further into this and hopefully will have a workaround for you soon

**Comment by Olivier-Vromans** [comment]
Yes I can see it, I just cloned it and executed `npm i` than I used `npm run dev` I clicked on test all and everything worked. I went over to Vercel and Deployed it there It build just fine:

```
[16:21:44.456] Cloning github.com/Olivier-Vromans/test (Branch: master, Commit: 074e7d8)
[16:21:45.377] Cloning completed: 920.383ms
[16:21:45.636] Looking up build cache...
[16:21:45.763] Build Cache not found
[16:21:45.814] Running "vercel build"
[16:21:46.491] Vercel CLI 28.4.12-05a80a4
[16:21:46.921] Installing dependencies...
[16:22:16.073] 
[16:22:16.073] added 362 packages in 29s
[16:22:16.074] 
[16:22:16.074] 101 packages are looking for funding
[16:22:16.074]   run `npm fund` for details
[16:22:16.097] Detected Next.js version: 13.0.0
[16:22:16.103] Detected `package-lock.json` generated by npm 7+...
[16:22:16.104] Running "npm run build"
[16:22:16.508] 
[16:22:16.508] > nextjs@0.1.0 build
[16:22:16.508] > next build
[16:22:16.508] 
[16:22:17.060] Attention: Next.js now collects completely anonymous telemetry regarding usage.
[16:22:17.060] This information is used to shape Next.js' roadmap and prioritize features.
[16:22:17.061] You can learn more, including how to opt-out if you'd not like to participate in this anonymous program, by visiting the following URL:
[16:22:17.061] https://nextjs.org/telemetry
[16:22:17.061] 
[16:22:17.202] info  - Linting and checking validity of types...
[16:22:17.484] info  - Creating an optimized production build...
[16:22:24.553] info  - Compiled successfully
[16:22:24.553] info  - Collecting page data...
[16:22:31.275] info  - Generating static pages (0/3)
[16:22:31.445] info  - Generating static pages (3/3)
[16:22:31.458] info  - Finalizing page optimization...
[16:22:31.462] 
[16:22:31.474] Route (pages)                              Size     First Load JS
[16:22:31.474] ┌ ○ /                                      1.94 kB        74.1 kB
[16:22:31.474] ├   └ css/ae0e3e027412e072.css             707 B
[16:22:31.474] ├   /_app                                  0 B            72.1 kB
[16:22:31.475] ├ ○ /404                                   181 B          72.3 kB
[16:22:31.475] ├ λ /api/execute                           0 B            72.1 kB
[16:22:31.475] └ λ /api/hello                             0 B            72.1 kB
[16:22:31.475] + First Load JS shared by all              72.4 kB
[16:22:31.475]   ├ chunks/framework-8c5acb0054140387.js   45.4 kB
[16:22:31.476]   ├ chunks/main-d2a793b6dc23a82a.js        25.7 kB
[16:22:31.476]   ├ chunks/pages/_app-b982fa09370b0414.js  284 B
[16:22:31.476]   ├ chunks/webpack-8fa1640cc84ba8fe.js     750 B
[16:22:31.476]   └ css/ab44ce7add5c3d11.css               247 B
[16:22:31.476] 
[16:22:31.476] λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
[16:22:31.476] ○  (Static)  automatically rendered as static HTML (uses no initial props)
[16:22:31.477] 
[16:22:32.875] Traced Next.js server files in: 1.231s
[16:22:33.934] Created all serverless functions in: 1.056s
[16:22:33.946] Collected static files (public/, static/, .next/static): 9.617ms
[16:22:35.171] Build Completed in /vercel/output [48s]
[16:22:40.208] Generated build outputs:
[16:22:40.208]  - Static files: 18
[16:22:40.208]  - Serverless Functions: 2
[16:22:40.208]  - Edge Functions: 0
[16:22:40.208] Serverless regions: Washington, D.C., USA
[16:22:40.208] Deployed outputs in 3s
[16:22:40.735] Build completed. Populating build cache...
[16:22:58.211] Uploading build cache [70.82 MB]...
[16:22:59.936] Build cache uploaded: 1.725s
[16:22:59.973] Done with "."
```

 but still this error:
```
[POST] /api/execute
16:23:01:41
No parameters
Run cypres for: 
Screenshot directories: 
2022-10-31T15:23:02.978Z	a1c98d64-0033-4643-8287-0329c1ca391d	ERROR	Error: No version of Cypress is installed in: /home/sbx_user1051/.cache/Cypress/10.11.0/Cypress
Please reinstall Cypress by running: cypress install
----------
Cypress executable not found at: /home/sbx_user1051/.cache/Cypress/10.11.0/Cypress/Cypress
----------
Platform: linux-x64 (Amazon Linux - 2 (Karoo))
Cypress Version: 10.11.0
    at /var/task/node_modules/cypress/lib/errors.js:348:17 {
  known: true
}
2022-10-31T15:23:12.033Z a1c98d64-0033-4643-8287-0329c1ca391d Task timed out after 10.01 seconds
```

**Comment by amehta265** [comment]
@Olivier-Vromans I don't understand where the error is stemming from. As you mentioned the build on Vercel was successful. So when is this error thrown? Is it thrown in Vercel or on your local machine?

Have you tried to install cypress from the folder you mentioned. Something like this: `./node_modules/vercel/.cache/Cypress install`

It is very difficult to debug this issue as I am unable to reproduce it on my end. As you have seen, in my reproduction I cant seem to emit the error you are getting despite producing an identical reproduction to yours. At this point, the only differences between my reproduction and yours would be our Vercel accounts and configurations on Vercel for this project and our systems (you are using a linux-x64 based Amazon Linux - 2 (Karoo) and I am using a Unix based macOS )

Is there any other information you can provide me that would help with a reproduction?

**Comment by Olivier-Vromans** [comment]
I get the error after the build on Vercel when I try to use the button that calls upon the execute.js file. Here in do i use the `cypress.run()` function. As for if I tried to install it from the cypress folder, yes I tried that and it was saying it was installed but when I called it afterwards it said No version of Cypress is installed. For my Vercel account I am using a Hobby account I don't know if that is the problem? Beside that I don't know what info I can give you that you can reproduce my problem

**Comment by amehta265** [comment]
Right now there doesn't seem to be enough information to reproduce the problem on our end. We'll have to close this issue until we can reproduce it. This does not mean that your issue is not happening - it just means that we do not have a path to move forward.
We have a [Cypress Discord server](https://on.cypress.io/chat) that could help you with this matter. Unfortunately I have done everything on my end to reproduce this but have not been able to.

Please open a new issue with a reproducible example and link to this issue. Here are some tips for providing a [Short, Self Contained, Correct, Example](http://sscce.org/) and our own [Troubleshooting Cypress](https://on.cypress.io/troubleshooting) guide.

**Comment by itailiors** [comment]
Hey @Olivier-Vromans, I have the exact same problem.
Did you solve it at the end?
If yes, remember how?

**Comment by atoulmet** [comment]
Hello!
I'm facing the same issue with the Cypress module API...
Is there anything new here?

**Comment by brian24C** [comment]
Hello!!
same issue here :(
